ark/utils/google_drive_utils.py

```diff
@@ -0,0 +1,555 @@
+
+    def clone(self, dest, overwrite=False, clear_dest=False):
+        """ Clones directory structure into provided destination
+
+        For a given path_string, e.g '/root/folderA/folderB', all files and subdirectories of
+        'folderB' are cloned into the destination.
+
+        Args:
+            dest (str):
+                A path location to contain cloned directory structure.  Must be a directory.
+            overwrite (bool):
+                If files are already present in dest and overwrite is False, no cloning is
+                performed.  Otherwise, the directory is overwriten with the cloned data.
+            clear_dest (bool):
+                If clear_dest is true, all contents of dest are removed before cloning. Default is
+                False.
+        """
+        if self._service_check() or self.fileID is None:
+            return
+
+        if clear_dest and os.path.exists(dest):
+            shutil.rmtree(dest)
+
+        if not os.path.exists(dest):
+            new_foldername = os.path.basename(dest)
+            dest = os.path.abspath(os.path.join(dest, os.pardir))
+            if not os.path.exists(dest):
+                raise FileNotFoundError(f"Could not resolve the path {dest}")
+
+            dest = os.path.join(dest, new_foldername)
+            os.mkdir(dest)
+
+        elif not overwrite:
+            print(
+                f"Warning: The path '{dest}' already exists. If you wish to overwrite/update files"
+                + " here, please pass the argument `overwrite=True`"
+            )
+            return
+
+        fname, fdata = self.get_name_and_data()
+        if fdata is None:
+            for filename in self.lsfiles():
+                filepath = self / filename
+                filepath.clone(dest, overwrite=True)
+            for dirname in self.lsdirs():
+                dirpath = self / dirname
+                dirpath.clone(os.path.join(dest, dirname), overwrite=overwrite)
+        else:
+            with open(os.path.join(dest, fname), 'wb') as f:
+                f.write(fdata.read())
+
+        return
+
+    def upload(self, src, overwrite=False):
+        """ Uploads contents and file structure of local directory to Drive.
+
+        For a given path_string, e.g '/root/folderA/folderB', the provided source file, as well as
+        all of its contents, will be added into 'folderB'.
+
+        Args:
+            src (str):
+                A path location to copy a directory structure/content from.  Doesn't need to be a
+                directory.
+            overwrite (bool):
+                If similarly named content exists on Drive, overwrite is False, no upload will take
+                place.  Otherwise, the content on the Drive folder is updated/overwritten.
+        """
+        if (
+            self._service_check()
+            or set(list(self.parent_id_map.keys())[1:]) != set(self.path_parents())
+        ):
+            return
+
+        if not os.path.exists(src):
```

Comment:
Might be helpful to provide some other error messages if the src path does not exist or if there are other failures.  There seem to be sufficient warnings for the overwrite cases and also for successful uploads, but it seems like this would return nothing if I incorrectly specified the local path containing the data I want to upload?